pkg/compactor/blocks_cleaner.go

```diff
@@ -150,7 +157,7 @@ func (c *BlocksCleaner) starting(ctx context.Context) error {
 }
 
 func (c *BlocksCleaner) ticker(ctx context.Context) error {
-	c.runCleanup(ctx)
+	go c.runCleanup(ctx)
```

Comment:
I'm not a big fan of this. It could lead to unbounded goroutines explosion if the function call gets stuck for any reason (or very slow). Also, I think it's way easier to reason about `runCleanup()` as non concurrency safe.

Reply:
I took a better look at https://github.com/grafana/dskit/pull/214 and my concern about goroutines explosition is waning. Reason is that if a tenant is stuck, then the next concurrent run won't run for that tenant (because it only runs for non in-flight tenants). Same if all tenants are stuck, then the next concurrent run won't run for any tenant at all.

Reply:
this runs once every 15m by default. to get 10K goroutines we'd need to run for 104 days and get every cleanup stuck.

One way to make runcleanup single-threaded is to use channels to pass jobs to the singleflight. Something like this b25b600f46ba58b664bea231d0d19f26e66e53e0. But we also need to keep track of the errors they return. Which starts sounding a lot like the ring.DoBatch and the item trackers there. Not sure if it's worth adding all that complexity for the occasional slow cleanup.

Reply:
posted that comment without seeing your last two...

> That being said, this is not safe. In the chain of calls done by runCleanup() there are blocks of code which is not concurrency safe.



is this the critical section that is not concurrency-safe?

https://github.com/grafana/mimir/blob/1925e2d758113363411a0a313fcf27ffc5d298fa/pkg/compactor/blocks_cleaner.go#L183-L203

Reply:
> is this the critical section that is not concurrency-safe?

Yes. I think the users scanner is safe, but some of the other code you linked isn't.

Reply:
> the bit in runCleanup requires that the cleanUser() returns when all cleanups are done, so we'd need cleanUser to be concurrency-safe too.... and runCleanup just calls cleanUser

I'm a bit lost. Aren't you saying the same thing I've said 👇 

> The actual per-tenant cleanup was already running concurrently, so it's expected to be concurrency safe. Please double check it and add comments to functions that are required to be concurrency safe.

Reply:
> Aren't you saying the same thing I've said
I don't think so. there is the code that updates metrics and logs that cleanup has completed. That code relies on the fact that all the tenant cleanups have completed (whether they ran concurrently or not). It checks if any of the tenant cleanups returned an error.

We cannot do `go c.singleFlight.ForEachNotInFlight(...)` because the function which would do `go c.singleFlight.ForEachNotInFlight(...)` needs to wait for all users to complete before returning. That function is `cleanUsers` and it returns to `runCleanup`, `runCleanup` in turn logs and updates metrics

Reply:
Right. I see what you mean. Then I think we should just make the function concurrency safe and add comments to the function to mention it.

Reply:
sounds good 👌

Reply:
posted my findings/work in a comment, will resolve this